paderbox/array/sparse.py

```diff
@@ -0,0 +1,701 @@
+            f'length={self.segment_length}: {self.array})'
+        )
+
+
+@dataclass
+class SparseArray:
+    """
+    A sparse array implementation using intervals to define non-zero regions.
+
+    >>> a = SparseArray(shape=(20, ))
+    >>> a
+    SparseArray(shape=(20,))
+    >>> a[5:10] = np.ones(5, dtype=int)
+    >>> a
+    SparseArray(
+      _SparseSegment(onset=5, length=5: [1 1 1 1 1]),
+      shape=(20,))
+    >>> np.asarray(a)
+    array([0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
+    >>> a[15:20] = 2
+    >>> a
+    SparseArray(
+      _SparseSegment(onset=5, length=5: [1 1 1 1 1])
+      _SparseSegment(onset=15, length=5: [2 2 2 2 2]),
+      shape=(20,))
+    >>> np.asarray(a)
+    array([0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 2, 2, 2, 2, 2])
+
+    Overlapping intervals are not allowed
+    >>> a[5:20] = 3
+    Traceback (most recent call last):
+      ...
+    ValueError: Overlap detected between
+      _SparseSegment(onset=5, length=5: [1 1 1 1 1])
+    and newly added
+      _SparseSegment(onset=5, length=15: [3 3 3 3 3 3 3 3 3 3 3 3 3 3 3])
+
+
+    dtype is inferred from segments:
+    >>> a.dtype
+    dtype('int64')
+    >>> a.pad_value, type(a.pad_value)
+    (0, <class 'numpy.int64'>)
+
+    Can't add segments with differing dtypes (yet)
+    >>> a[:3] = np.ones(3, dtype=np.float32)
+    Traceback (most recent call last):
+      ...
+    TypeError: Type mismatch: SparseArray has dtype int64, but assigned array has dtype float32
+
+    Adding multiple SparseArray returns a SparseArray
+    >>> b = SparseArray(a.shape)
+    >>> b[10:15] = 8
+    >>> a + b
+    SparseArray(
+      _SparseSegment(onset=5, length=5: [1 1 1 1 1])
+      _SparseSegment(onset=10, length=5: [8 8 8 8 8])
+      _SparseSegment(onset=15, length=5: [2 2 2 2 2]),
+      shape=(20,))
+    >>> np.asarray(a + b)
+    array([0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 8, 8, 8, 8, 8, 2, 2, 2, 2, 2])
+
+    Arithmetic works with numpy arrays
+    >>> np.ones(20, dtype=int) * a
+    array([0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 2, 2, 2, 2, 2])
+    >>> np.ones(20, dtype=int) + a
+    array([1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 3, 3, 3, 3, 3])
+    >>> a + np.ones(20, dtype=int)
+    array([1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 3, 3, 3, 3, 3])
+    >>> c = np.ones(20, dtype=int)
+    >>> c += a
+    >>> c
+    array([1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 3, 3, 3, 3, 3])
+
+    """
+    # Public constructor interface
+    shape: tuple
+
+    # Private constructor arguments: Don't use _segments directly.
+    # This is required for pt.data.batch.example_to_device to work
+    _segments: list = field(default_factory=list)
+
+    # Other instance attributes
+    pad_value = 0.
+    _is_torch = None
+    dtype = None
+
+    def __post_init__(self):
+        self.shape = _normalize_shape(self.shape)
+        self._update_dtype(self.dtype)
+
+    @classmethod
+    def from_array_and_offset(cls, array, offset, shape):
```

Comment:
This is confusing. We should try to avoid the name `offset`, because it is sometimes the start point and sometimes the endpoint. `onset` has a unique definition.